mediaapi/routing/preview_url.go

```diff
@@ -0,0 +1,400 @@
+	OgType          string `json:"og:type"`
+	OgTitle         string `json:"og:title"`
+	OgUrl           string `json:"og:url"`
+	OgDescription   string `json:"og:description"`
+}
+
+// PreviewUrl implements GET /preview_url.
+// Current implementation gets the url, parses the meta tags to obtain OG
+// data and returns a JSONResponse for the client to process.
+// TODO: avoid this endpoint in encrypted rooms
+func PreviewUrl(
+	req *http.Request, cfg *config.MediaAPI, db storage.Database) util.JSONResponse {
+
+	// parsing the request
+	r, err := parseRequest(req)
+	if err != nil {
+		return *err
+	}
+
+	// downloading metadata for the url
+	mediaInfo, err := r.downloadUrl(cfg, req.Context(), db)
+	if err != nil {
+		return *err
+	}
+
+	// preparing the response
+	response, err := mediaInfo.prepareResponse(cfg)
+	if err != nil {
+		return *err
+	}
+
+	return util.JSONResponse{
+		Code: http.StatusOK,
+		JSON: response,
+	}
+}
+
+// parseRequest parses the incoming preview request to extract the url and ts.
+// Returns either a parsed PreviewUrlRequest or error formatted as
+// util.JSONResponse
+func parseRequest(req *http.Request) (*PreviewUrlRequest, *util.JSONResponse) {
+
+	// get the url to be previewed from the request
+	urlToPreview := req.URL.Query().Get("url")
+	if len(urlToPreview) == 0 {
+		return nil, &util.JSONResponse{
+			Code: http.StatusBadRequest,
+			JSON: jsonerror.MissingArgument("Missing URL"),
+		}
+	}
+
+	// parse the url
+	parsedUrl, err := url.Parse(urlToPreview)
+	if err != nil {
+		return nil, &util.JSONResponse{
+			Code: http.StatusBadRequest,
+			JSON: jsonerror.InvalidArgumentValue(
+				"Unable to parse url " + err.Error()),
+		}
+	}
+
+	// get the ts if provided in the request
+	if tsStr := req.URL.Query().Get("ts"); len(tsStr) > 0 {
+		ts, err := strconv.ParseInt(tsStr, 10, 64)
+		if err != nil {
+			return nil, &util.JSONResponse{
+				Code: http.StatusBadRequest,
+				JSON: jsonerror.InvalidArgumentValue(
+					"Couldn't parse 'ts' to a valid integer" + err.Error()),
+			}
+		}
+
+		request := &PreviewUrlRequest{
+			url: *parsedUrl,
+			// Convert timestamp to ms
+			ts: types.UnixMs(ts / 1000000),
+		}
+
+		return request, nil
+	}
+
+	// set ts to current time if none provided
+	ts := time.Now().UnixNano()
+
+	request := &PreviewUrlRequest{
+		url: *parsedUrl,
+		ts:  types.UnixMs(ts / 1000000),
+	}
+
+	return request, nil
+}
+
+// downloadUrl downloads the url and saves the response body.
+// Returns either mediaInfo (metadata about the file to preview) or error
+// formatted as util.JSONResponse.
+// Current implementation for saving response is heavily similar to /upload,
+// need to work on this.
+// TODO: better logging
+func (previewReq *PreviewUrlRequest) downloadUrl(cfg *config.MediaAPI, ctx context.Context, db storage.Database) (*mediaInfo, *util.JSONResponse) {
+
+	urlString := previewReq.url.String()
+
+	// Get the URL
+	response, err := http.Get(urlString)
```

Comment:
Don't use the default client as it has no timeout, so malicious preview URLs could just not return anything and consume a file descriptor on the server. Use a client with a sensible timeout please e.g `&http.Client{Timeout:30 * time.Second}`